### DISCUSSION TITLE
c7n Release Checklist

### DISCUSSION THREAD

**Discussion by castrojo** [cloud-custodian/cloud-custodian#7509] [General]
# Release Checklist Notes

This document is only useful for making a cloud custodian
release. We're in the process of automating this but posting it here
for the record and for people who want to play.

## Prerequisites / Assumptions

**Use Python 3.10 or higher** - c7n-left requires 3.10+ to install, and to release

* Releases are built and published using the `main` branch
* The release process documented here should come _after_ a "release preparation" pull request gets merged. That pull request should:
  * Increment package versions (via `make pkg-increment`)
  * Update dependencies (via `make pkg-rebase`)
  * Resolve any test breakages related to those dependency updates

_For reference, [here](https://github.com/cloud-custodian/cloud-custodian/pull/7492) is an example pull request preparing for release `0.9.17.0`_

You'll need [GitHub CLI](https://cli.github.com) and `twine` installed as well.

## Run Functional Tests

before releasing we run functional tests to ensure trunk is suitable for release:

    gh workflow run functional.yaml

on success it will notify into the slack channel `#github-ci`

## Release artifact creation

Producing wheels and changelog has been automated via GitHub actions, to create release artifacts:

    gh workflow run release.yml

This will create wheels, upload them to a private pypi (code artifact), test wheel installation, test basic usage, and generate a changelog.

Download the release artifacts into a clean working copy:

    make release-get-artifacts

Open the Changelog (`release.md`) in a text editor and adjust accordingly so that it makes sense for humans. 
    
## Create a new release on c7n repo

Paste in changelog when making a release:

    cat release.md | pbcopy # OSX-only
    cat release.md | xsel -c # Linux/X11 .. for Wayland see wl-clipboard

Create a new tag

Github [documentation on managing releases](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository)

# Publish Wheel

    make pkg-publish-wheel PKG_REPO=prodpypi

## Announce

Publish announcement to the [c7n mailing list] and [slack, gitter channel].

**Comment by darrendao** [comment]
Also need to have latest pip version installed
```
pip install --upgrade pip
```

**Reply by ajkerrigan** [reply]
Thanks @darrendao , edited the discussion to include pip upgrades and some context around release prep expectations.

**Comment by thisisshi** [comment]
Note that 3.10 is required to release c7n-left

**Comment by kapilt** [comment]
updated to reflect use of poetry to publish, and running functional tests prior to release.

**Comment by kapilt** [comment]
updated for new automation on building artifacts/wheels.
